src/System.Device.Gpio/System/Device/Gpio/Drivers/Libgpiod/LibGpiodDriverFactory.cs

```diff
@@ -33,83 +33,95 @@ internal static class LibGpiodDriverFactory
+
+    /// <summary>
+    /// The value set by DOTNET_IOT_LIBGPIOD_DRIVER_VERSION. Null when not set.
+    /// </summary>
+    private string? DriverVersionEnvVarValue => _driverVersionEnvVarValue.Value;
+    private readonly Lazy<string?> _driverVersionEnvVarValue;
+
+    /// <summary>
+    /// The driver version that was resolved based on the value of DOTNET_IOT_LIBGPIOD_DRIVER_VERSION. Null when env var not set.
+    /// </summary>
+    private LibGpiodDriverVersion? DriverVersionSetByEnvVar => _driverVersionSetByEnvVar.Value;
+    private readonly Lazy<LibGpiodDriverVersion?> _driverVersionSetByEnvVar;
 
     /// <remarks>
     /// Driver version that is picked by this factory when no version is explicitly specified. Null when libpgiod is not installed.
     /// </remarks>
-    private static readonly LibGpiodDriverVersion? _automaticallySelectedDriverVersion;
+    private LibGpiodDriverVersion? AutomaticallySelectedDriverVersion => _automaticallySelectedDriverVersion.Value;
+    private readonly Lazy<LibGpiodDriverVersion?> _automaticallySelectedDriverVersion;
 
-    private static readonly string[] _librarySearchPaths = { "/lib", "/usr/lib", "/usr/local/lib" }; // Based on Linux FHS standard
+    private readonly Lazy<LibGpiodDriverVersion[]> _driverCandidates;
 
-    public static readonly LibGpiodDriverVersion[] DriverCandidates;
+    /// <summary>
+    /// Collection of installed libgpiod libraries (their file name).
+    /// </summary>
+    private IEnumerable<string> InstalledLibraries => _installedLibraries.Value;
+    private readonly Lazy<IEnumerable<string>> _installedLibraries;
 
-    static LibGpiodDriverFactory()
+    private LibGpiodDriverFactory()
     {
-        var installedLibraries = GetInstalledLibraries();
-
-        DriverCandidates = installedLibraries.Where(installedVersion => _libraryToDriverVersionMap.ContainsKey(installedVersion))
-                                              .Select(installedVersion => _libraryToDriverVersionMap[installedVersion]).ToArray();
+        _installedLibraries = new Lazy<IEnumerable<string>>(GetInstalledLibraries, LazyThreadSafetyMode.PublicationOnly);
+        _driverCandidates = new Lazy<LibGpiodDriverVersion[]>(GetDriverCandidates, LazyThreadSafetyMode.PublicationOnly);
+        _automaticallySelectedDriverVersion = new Lazy<LibGpiodDriverVersion?>(GetAutomaticallySelectedDriverVersion, LazyThreadSafetyMode.PublicationOnly);
+        _driverVersionEnvVarValue = new Lazy<string?>(GetDriverVersionEnvVarValue, LazyThreadSafetyMode.PublicationOnly);
+        _driverVersionSetByEnvVar = new Lazy<LibGpiodDriverVersion?>(GetDriverVersionSetByEnvVar, LazyThreadSafetyMode.PublicationOnly);
+    }
 
-        _automaticallySelectedDriverVersion = DriverCandidates.Any() ? DriverCandidates.Max() : null;
+    /// <summary>
+    /// Singleton instance.
+    /// </summary>
+    public static LibGpiodDriverFactory Instance { get; } = new();
```

Comment:
This singleton pattern is not thread safe. Not sure it would matter here, but better use one of the templates I added to the last post.

Reply:
I think it is thread safe as it is similar to the following which is thread safe:
```
private static LibGpiodDriverFactory _instance = new();
public LibGpiodDriverFactory Instance => _instance;
```

The static fields are initialized lazily, as soon as anything related to that class is ever used.
Other singleton patterns are not necessarily thread safe, but the static fields initialization is guaranteed to be atomically initialized from the CLR.

Reply:
True, you're right. But it's not lazy, as indicated in some other comment.

Reply:
It is, until you use the `LibGpiodDriverFactory` for any reason.
As soon as the `JIT` sees that type in a method, it triggers its initialization and therefore the instance is created.

There are other, slightly more complex, implementations of the Singleton pattern which are thread-safe and use weak initialization. But I don´t believe it is worth the effort for this type.

Reply:
I believe we are saying the same things:
- The initialization is lazy because it is done only after the type is used (which is good IMO)
- That specific implementation of the singleton pattern is thread safe, so there is no need to use `lock`.

As soon as a different implementation is chosen (static method creating the instance), the use of `lock` is needed to guarantee the thread safety.

Also, I agree with Patrick that the goal here is to move away from the static constructor because, should a `TypeInitializationException` occur inside it, the exception argument would not provide enough clear clues to understand the culprit and, of course, will make the entire app immediately fail.

Reply:
I risk to go OT now, but let's look at the following snippet run with both  .NET Framework 4.7 and .NET 8.
This uses only "old" C# features to make it run with .NET Framework as well:
```csharp
using System;
public class C {
    private static C _instance = new C();
    public static C Instance { get { return _instance; } }

    public static void C1() { Console.WriteLine("C1"); }
    private C() { Console.WriteLine("Instance created"); }
}

public class Program
{
    public static void Main()
    {
        Console.WriteLine("Main");
        var p = new Program();
        p.M1();
        p.M2();
    }
    
    public void M1() { Console.WriteLine("M1"); }
    public void M2()
    {
        Console.WriteLine("M2 {0}", typeof(C));
        C.C1();
        C.Instance.ToString();
    }   
}
``` 

.NET Framework (regardless the version 4.7 or 4.8)
```
Main
M1
M2 C
Instance created
C1
```

.NET (regardless the version) >= 3.x:
```
Main
M1
M2 C
C1
Instance created
``` 

The point is that this implementation is always lazy. In .NET (core) the instance is not created until you really want to access the instance method.
If you just stick with static members, the instance is never created at all.

Reply:
We are on the same page. I did not know this subtle difference between the dotnet versions, although it doesn't really matter here because `Instance` is the only public static member, so regardless of whether `get { ... }` or static constructor is used, only accessing `Instance`  will cause initialization (or maybe I am missing something?).
To summarize a bit: this factory should perform the initialization logic 1. only once 2. lazy and 3. thread safe. The current implementation utilizes the `Lazy` type to achieve 2. 3. and `static` for 1. When referencing `Instance`, the (automatically generated) static constructor gets called that initializes the `Lazy` members. This will not throw and consequently `TypeInitializationException` is avoided.
When we compare this solution with one that uses `get { ... }` and a lock, I only see performance being the last factor left, but I don't think it's worth the effort going there. This class will most likely not be used in hot spots, unless someone has the idea to create millions of  driver instances for some reason 😃

Reply:
I'm not sure about the thread safety of the getter initialization. It probably is, I would rather prefer the following, which is a one-liner having all the requirements and leave no doubts :-)
```csharp
public static readonly LibGpiodDriverFactory Instance = new();
```

Reply:
I might as well just use the other approach. I find it's readability better :)